File: python/pull_acs_ice.py
#!/usr/bin/env python3
"""
pull_acs_ice.py — Pull county-level ACS 5-year data and compute the
Index of Concentration at the Extremes for racialized economic segregation,
ICE(race+income), following Krieger et al. (AJPH 2016).

WHY YOU RUN THIS LOCALLY: the Cowork sandbox cannot reach api.census.gov.
Run this on your machine; it writes county_ice.csv into data/raw/ where the
pipeline picks it up automatically.

Definition (household-based, fully reproducible):
  A_i = # White non-Hispanic households with income >= $100,000   (privileged extreme)
  P_i = # households of color (all races minus White NH) with income < $25,000 (deprived extreme)
  T_i = total households with income data
  ICE_i = (A_i - P_i) / T_i        in [-1, +1];  +1 = all privileged, -1 = all deprived

Tables: B19001 (household income, all) and B19001H (household income, White alone not Hispanic).
Vintage: ACS 2018-2022 5-year (year=2022) to match the State Cancer Profiles
'Latest 5-year average' incidence window. Change ACS_YEAR if needed.

Usage:
  pip install requests
  export CENSUS_API_KEY=xxxx     # optional; get a free key at https://api.census.gov/data/key_signup.html
  python pull_acs_ice.py
"""
import os, sys, csv, json, urllib.request, urllib.parse

ACS_YEAR = 2022
BASE = f"https://api.census.gov/data/{ACS_YEAR}/acs/acs5"
KEY = os.environ.get("CENSUS_API_KEY", "").strip()
OUT = os.path.join(os.path.dirname(__file__), "..", "data", "raw", "county_ice.csv")

# B19001 income brackets: _002.._017. <25k = 002,003,004,005 ; >=100k = 014,015,016,017
LOW  = ["002", "003", "004", "005"]            # <10k,10-15,15-20,20-25
HIGH = ["014", "015", "016", "017"]            # 100-125,125-150,150-200,>=200k

def fetch(table):
    cols = [f"{table}_{i:03d}E" for i in range(1, 18)]  # _001 total .. _017
    params = {"get": "NAME," + ",".join(cols), "for": "county:*"}
    if KEY:
        params["key"] = KEY
    url = BASE + "?" + urllib.parse.urlencode(params)
    with urllib.request.urlopen(url, timeout=120) as r:
        data = json.load(r)
    head, *rows = data
    idx = {h: k for k, h in enumerate(head)}
    out = {}
    for row in rows:
        fips = row[idx["state"]] + row[idx["county"]]
        out[fips] = {c: (int(row[idx[c]]) if row[idx[c]] not in (None, "", "-") else 0) for c in cols}
        out[fips]["NAME"] = row[idx["NAME"]]
    return out, table

def s(d, table, codes):
    return sum(d[f"{table}_{c}E"] for c in codes)

def fetch_pop():
    """Total county population (B01003_001E) for population-weighting."""
    params = {"get": "B01003_001E", "for": "county:*"}
    if KEY:
        params["key"] = KEY
    url = BASE + "?" + urllib.parse.urlencode(params)
    with urllib.request.urlopen(url, timeout=120) as r:
        data = json.load(r)
    head, *rows = data
    idx = {h: k for k, h in enumerate(head)}
    out = {}
    for row in rows:
        fips = row[idx["state"]] + row[idx["county"]]
        v = row[idx["B01003_001E"]]
        out[fips] = int(v) if v not in (None, "", "-") else 0
    return out

def main():
    print(f"Pulling ACS {ACS_YEAR} 5-year B19001 + B19001H (county) ...", flush=True)
    allinc, _ = fetch("B19001")
    whinc, _  = fetch("B19001H")
    pop = fetch_pop()
    rows = []
    for fips, d in allinc.items():
        total = d["B19001_001E"]
        if total == 0 or fips not in whinc:
            continue
        w = whinc[fips]
        A = s(w, "B19001H", HIGH)                       # white NH, >=100k
        low_all = s(d, "B19001", LOW)                   # all races, <25k
        low_wh  = s(w, "B19001H", LOW)                  # white NH, <25k
        P = max(low_all - low_wh, 0)                    # households of color, <25k
        ice = (A - P) / total
        rows.append({"fips": fips, "county_name": d["NAME"], "pop_total": pop.get(fips, 0),
                     "hh_total": total, "A_white_highinc": A, "P_color_lowinc": P,
                     "ice_raceinc": round(ice, 5)})
    rows.sort(key=lambda r: r["fips"])
    os.makedirs(os.path.dirname(OUT), exist_ok=True)
    with open(OUT, "w", newline="") as f:
        w = csv.DictWriter(f, fieldnames=list(rows[0].keys()))
        w.writeheader(); w.writerows(rows)
    print(f"Wrote {len(rows)} counties -> {os.path.abspath(OUT)}")
    print("ICE range:", min(r['ice_raceinc'] for r in rows), "to", max(r['ice_raceinc'] for r in rows))

if __name__ == "__main__":
    try:
        main()
    except Exception as e:
        print("ERROR:", e, file=sys.stderr)
        print("If this is a network/key issue, get a free key at "
              "https://api.census.gov/data/key_signup.html and set CENSUS_API_KEY.", file=sys.stderr)
        sys.exit(1)

File: R/pull_acs_ice.R
#!/usr/bin/env Rscript
# pull_acs_ice.R — county ICE(race+income) + total population from ACS 5-year,
# using tidycensus. Run this LOCALLY (the Cowork sandbox cannot reach the Census API).
# Writes ../data/raw/county_ice.csv for the pipeline to pick up.
#
# ICE(race+income), Krieger et al. (AJPH 2016), household-based:
#   A = White non-Hispanic households with income >= $100,000      (B19001H 014:017)
#   P = households of color (all - White NH) with income < $25,000  (B19001 002:005 - B19001H 002:005)
#   T = total households (B19001_001);  ICE = (A - P) / T  in [-1, 1]; lower = more deprived
#
# Setup (once):
#   install.packages(c("tidycensus","dplyr","tidyr","readr"))
#   tidycensus::census_api_key("YOUR_KEY", install = TRUE)   # free: https://api.census.gov/data/key_signup.html
# Then:
#   Rscript pull_acs_ice.R

suppressPackageStartupMessages({
  library(tidycensus); library(dplyr); library(tidyr); library(readr)
})

ACS_YEAR <- 2022   # ACS 2018-2022 5-year, matches 'Latest 5-year average' cancer window
# Robust to working directory: works whether wd is the project root or the R/ folder.
find_out <- function() {
  for (d in c("data/raw", "../data/raw", file.path(getwd(), "data/raw"))) {
    if (dir.exists(d)) return(file.path(d, "county_ice.csv"))
  }
  dir.create("data/raw", recursive = TRUE, showWarnings = FALSE)
  "data/raw/county_ice.csv"
}
out_path <- find_out()

low_codes  <- sprintf("%03d", 2:5)    # <10k,10-15,15-20,20-25  -> <25k
high_codes <- sprintf("%03d", 14:17)  # 100-125,125-150,150-200,>=200k -> >=100k

get_tbl <- function(tbl) {
  get_acs(geography = "county", table = tbl, year = ACS_YEAR,
          survey = "acs5", output = "wide", cache_table = TRUE) %>%
    select(GEOID, NAME, ends_with("E"))
}

message("Pulling ACS ", ACS_YEAR, " 5-year: B19001, B19001H, B01003 (county) ...")
all_inc <- get_tbl("B19001")
wh_inc  <- get_tbl("B19001H")
pop     <- get_acs(geography = "county", variables = "B01003_001",
                   year = ACS_YEAR, survey = "acs5", output = "wide") %>%
  transmute(GEOID, pop_total = B01003_001E)

sum_codes <- function(df, tbl, codes) rowSums(df[, paste0(tbl, "_", codes, "E"), drop = FALSE], na.rm = TRUE)

res <- all_inc %>%
  transmute(fips = GEOID, county_name = NAME,
            hh_total = B19001_001E,
            low_all  = sum_codes(all_inc, "B19001", low_codes)) %>%
  left_join(wh_inc %>% transmute(fips = GEOID,
              A_white_highinc = sum_codes(wh_inc, "B19001H", high_codes),
              low_wh          = sum_codes(wh_inc, "B19001H", low_codes)), by = "fips") %>%
  left_join(pop, by = c("fips" = "GEOID")) %>%
  mutate(P_color_lowinc = pmax(low_all - low_wh, 0),
         ice_raceinc    = round((A_white_highinc - P_color_lowinc) / hh_total, 5)) %>%
  filter(hh_total > 0) %>%
  select(fips, county_name, pop_total, hh_total, A_white_highinc, P_color_lowinc, ice_raceinc) %>%
  arrange(fips)

setwd("~/Library/CloudStorage/GoogleDrive-kgahan.srf.aiims@gmail.com/My Drive/suppression-bias")
tidycensus::census_api_key("a6c65f8a5107f9105f56ca8f674de3866d64441e", install = TRUE, overwrite = TRUE)
readRenviron("~/.Renviron")
source("R/pull_acs_ice.R")

dir.create(dirname(out_path), showWarnings = FALSE, recursive = TRUE)
write_csv(res, out_path)
message("Wrote ", nrow(res), " counties -> ", normalizePath(out_path, mustWork = FALSE))
message("ICE range: ", min(res$ice_raceinc), " to ", max(res$ice_raceinc))
